Added support for cursor operations and a color helper to IO.ANSI.

Calamitous opened the pull request.

josevalim commented on lib/elixir/lib/io/ansi.ex:
```diff
@@ -46,6 +52,14 @@ defmodule IO.ANSI do
+  Valid values for each color value are in the range 0 to 5.
+  """
+  @spec rgb216(0..5, 0..5, 0..5) :: String.t
+  def rgb216(r, g, b) when valid_rgb216(r, g, b), do: 16 + (36 * r) + (6 * g) + b
```
I wonder if we should just make it `color/3` or `rgb_color/3` so you don't need to call two functions.

Calamitous replied:
Yeah, that makes a lot more sense.  I could make color/3 that calls color/1, so either method is still available.  I'll make that tweak and re-submit, thanks for the input!

josevalim commented:
I am not comfortable with adding all the cursor functions for now. Can we just add `color/3` for now? The remaining ones can be handled on your own project, thank you. :)

Calamitous commented:
I've change the color helper to color/3, and added color_background/3 with the same functionality.

I've also removed the cursor operations from this PR.  If I may ask, was it a code issue that made you uncomfortable with the cursor functions, or do you feel that those don't belong in the IO.ANSI lib?

lexmag commented on lib/elixir/test/elixir/io/ansi_test.exs:
```diff
@@ -120,4 +132,17 @@ defmodule IO.ANSITest do
+      IO.ANSI.color_background(256)
+    end
+  end
+
```
Needless line :scissors:.

lexmag commented on lib/elixir/lib/io/ansi.ex:
```diff
@@ -29,6 +29,12 @@ defmodule IO.ANSI do
   @typep ansilist :: maybe_improper_list(char() | ansicode() | binary() | ansilist(), binary() | ansicode() | [])
   @type  ansidata :: ansilist() | ansicode() | binary()
 
+  defmacro valid_rgb216(r, g, b) do
```
Maybe we should move this check into `color` and `color_background` definitions?
It is simple enough to be defined directly in the user code, furthermore, people will need to `require IO.ANSI`.

Calamitous replied:
That's a good point.  I originally wrote the validation as a macro to use it as a guard clause, but looking at it again, the logic is simple enough that a macro is not really justified.  I can fix that up and submit another PR.

As a side question: I see that the defsequence macro is defined in a separate module and required in IO.ANSI-- is this what keeps users from having to `require IO.ANSI`?

Calamitous replied:
Ah, nm... I see you've already fixed it... thanks!

josevalim commented:
> I've also removed the cursor operations from this PR. If I may ask, was it a code issue that made you uncomfortable with the cursor functions, or do you feel that those don't belong in the IO.ANSI lib?

I don't feel it belongs just yet. IO.ANSI was supposed to be mostly about colors and formatting, I am a bit wary of taking it too far away from that. :)

Calamitous commented:
> I don't feel it belongs just yet. IO.ANSI was supposed to be mostly about colors and formatting, I am a bit wary of taking it too far away from that. :)

Fair enough.  Feel free to give me a shout if you ever want the cursor functions back in and I'll be happy to take care of that.  :)

eksperimental commented:
Since valid_rgb216 is a public macro and is allowed in guards, shouldnt it be called is_valid_rgb216  Or ar least valid_rgb216? Since it returns boolean

josevalim commented:
The macro was removed in a later commit.

eksperimental commented:
All good then, thank you jose
